### PR TITLE
Avoid encoding data-amp-addparams 

### LINKED ISSUE #14958: Appending a Query Parameter with a default value to the href in links does not work
## What's the issue?

We are trying to use variable substitution in links as documented [here](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md#query-parameter) and [here](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md#query-parameter).

Here is an example of the HTML we are using:

`<a href="//example.com/" data-amp-addparams="SRC=QUERY_PARAM(SRC,XXXX)" data-amp-replace="QUERY_PARAM">Go to my site</a>`

Clicking on the above link we get `//example.com/?SRC=(SRC,XXXX)` instead of linking to `//example.com/?SRC=XXXX`

There is nowhere that explicitly states in the AMP documentation that using `QUERY_PARAM` with a default value in combination with `data-amp-addparams` will give the desired outcome.  However, there is nowhere explicitly stating that it will not.  If there is a valid reason for why this can't be done it may be best to highlight this in the docs.

## How do we reproduce the issue?

Here is a link to the test case - http://jsbin.com/pesevax/edit?html,output

4 scenarios are outlined in the jsbin with the final scenario highlighting the issue.

## What browsers are affected?

Seems to affect all browsers. We have observed this behaviour in Chrome, Safari, Firefox and Microsoft Edge.

## Which AMP version is affected?

Version 1524089272632 but I believe this was always an issue.

### PR DESCRIPTION
**Pull request by micajuine-ho**
Close #14958.

Expands `data-amp-params` if they are global macros and they have been declared in `data-amp-replace` and it is an allowed origin.

### CODE REVIEW

**Comment by zhouyx** [comment]
Thanks for the fix. However one thing I want to point out is that `data-amp-addparams` can be used to append variables that do need to be encoded. 

A proper fix to not break existing users is to find all macros, not encode them, but encode the rest of the params. Or expand the params immediately. Somehow similar to #22379

**Comment by micajuine-ho** [comment]
Discussed offline, will go with solution where we resolve macros before encoding them. 
Final process will look like this:
- Find macros
- Resolve them
- Encode everything else
- Resolve macros found concatenated url

Will wait for #26271 to merge before implementing, since it handles the same problem in a different place

File: src/service/url-replacements-impl.js [hunk 1]
```diff
@@ -1077,11 +1077,15 @@ export class UrlReplacements {
     if (element[ORIGINAL_HREF_PROPERTY] == null) {
       element[ORIGINAL_HREF_PROPERTY] = href;
     }
+
+    const isAllowedOrigin = this.isAllowedOrigin_(url);
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: Let's return early if the origin is not allowed.

File: src/service/url-replacements-impl.js [hunk 2]
```diff
@@ -1116,18 +1120,27 @@ export class UrlReplacements {
...
 
-    if (whitelist) {
-      href = this.expandUrlSync(
-        href,
-        /* opt_bindings */ undefined,
-        /* opt_collectVars */ undefined,
-        /* opt_whitelist */ whitelist
-      );
-    }
+    href = this.expandSyncIfWhitelist_(href, whitelist);
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: Do we need to expand again?

File: src/service/url-replacements-impl.js [hunk 3]
```diff
@@ -1116,18 +1120,27 @@ export class UrlReplacements {
...
 
     return (element.href = href);
   }
 
+  /**
+   * @param {string} href
+   * @param {!Object<string, boolean>|undefined} whitelist
+   * @return {string}
+   */
+  expandSyncIfWhitelist_(href, whitelist) {
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: nit: switch to allowedList please : )

File: test/unit/test-url-replacements.js [hunk 4]
```diff
@@ -1931,6 +1931,21 @@ describes.sandboxed('UrlReplacements', {}, env => {
           );
         });
 
+        it('should concatenate and expand additional params w/ whitelist', () => {
```
Comment (zhouyx) [inline test/unit/test-url-replacements.js]: Could you add another param that does need encoding and make sure we still encode the variable. Thanks.

File: src/service/url-replacements-impl.js (hunk 1 above)
Comment (micajuine-ho) [reply src/service/url-replacements-impl.js]: Sounds good

File: src/service/url-replacements-impl.js (hunk 2 above)
Comment (micajuine-ho) [reply src/service/url-replacements-impl.js]: Yes, the first expansion is for the `data-amp-addparams` if they exist. This expansion is for the `href` as a whole.

Comment (zhouyx) [reply src/service/url-replacements-impl.js]: Make sense. Thanks for the clarification.

File: test/unit/test-url-replacements.js (hunk 4 above)
Comment (micajuine-ho) [reply test/unit/test-url-replacements.js]: `CLIENT_ID` is not [defined in the `data-amp-replace`](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md#appending-parameters-to-the-href) attribute so it's parameters should be encoded.

I added more parameters to `CLIENT_ID` to show that it is encoded

Comment (zhouyx) [reply test/unit/test-url-replacements.js]: I meant  something like `data-amp-addparams=encode,test`

File: src/service/url-replacements-impl.js [hunk 5]
```diff
@@ -1095,6 +1092,13 @@ export class UrlReplacements {
       return (element.href = href);
     }
 
+    if (additionalUrlParameters) {
+      additionalUrlParameters = isAllowedOrigin
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: No need to check for `isAllowedOrigin` again.

File: src/service/url-replacements-impl.js [hunk 6]
```diff
@@ -1116,18 +1120,27 @@ export class UrlReplacements {
...
 
     return (element.href = href);
   }
 
+  /**
+   * @param {string} href
+   * @param {!Object<string, boolean>|undefined} whitelist
+   * @return {string}
+   */
+  expandSyncIfAllowedList_(href, whitelist) {
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: nit: `s/whitelist/allowedList`

File: src/service/url-replacements-impl.js [hunk 7]
```diff
@@ -1072,6 +1068,14 @@ export class UrlReplacements {
       return (element.href = href);
     }
 
+    if (additionalUrlParameters) {
+      additionalUrlParameters = this.expandSyncIfAllowedList_(
+        additionalUrlParameters,
+        whitelist
+      );
+      href = addParamsToUrl(href, parseQueryString(additionalUrlParameters));
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: you don't need this.

Comment (micajuine-ho) [reply src/service/url-replacements-impl.js]: Can you clarify?

Comment (zhouyx) [reply src/service/url-replacements-impl.js]: See line 1097

Comment (zhouyx) [reply src/service/url-replacements-impl.js]: ignore me

File: src/service/url-replacements-impl.js (hunk 1 above)
Comment (micajuine-ho) [reply src/service/url-replacements-impl.js]: Oops, looks like we can't resolve early. [Some of the tests](https://github.com/ampproject/amphtml/blob/master/test/unit/test-url-replacements.js#L1902-L1943) indicate that even if it's not an allowed origin, the `data-amp-addparams` should still be appended.

Comment (zhouyx) [reply src/service/url-replacements-impl.js]: 👍 Good call. LGTM

**Pull request merged**